In [3]:
%env METIS_DLL=/home/juan/Descargas/metis-5.0/build/Linux-x86_64/libmetis/libmetis.so

env: METIS_DLL=/home/juan/Descargas/metis-5.0/build/Linux-x86_64/libmetis/libmetis.so


In [1]:
import scipy

scipy.version.version

'1.5.4'

In [2]:
from scipy.sparse import lil_matrix, csr_matrix
import numpy as np
#import metis
from clustering_evaluation import computeEvaluationMeasures, MEASURES
import glob
from tabulate import tabulate
from tqdm import tqdm


def _loadSparseCSRFromDisk(fullPath):
    spD = None
    with open(fullPath, 'rb') as f:
        data = np.load(f)
        indices = np.load(f)
        indptr = np.load(f)
        spD = csr_matrix((data, indices, indptr))
        # convert this nxk matrix into a similarity.
    return spD

def convert_dense_csv_to_csr_fmt(denseFn, spM):
    csrOut = "{0}_csr.npy".format(denseFn[: -(len(denseFn.split('.')[-1])+1)]) # replaces the file extension.
    print(csrOut)
    #M  = np.loadtxt(denseFn, delimiter=',')
    #spM = csr_matrix(M)
    with open(csrOut, 'wb') as f:
        np.save(f, spM.data)
        np.save(f, spM.indices)
        np.save(f, spM.indptr)
    """
    When loading
    >> with open('test.npy', 'rb') as f:
    >>     a = np.load(f)
    >>     b = np.load(f)
    >> csr_matrix((data, indices, indptr))
    """
    


def getLabelsFromFile(fullPath):
    v = np.loadtxt(fullPath, delimiter=',', dtype=np.int)
    return v

def detectPostfixesInDir(fullPath, dirName):
    fnLst = glob.glob('{0}/{1}/*labels_*.out'.format(fullPath, dirName))
    # labels_NCF_DS_
    postFixes = [f.split("/")[-1].replace('labels_NCF_DS_','').replace('.out','') for f in fnLst]
    return postFixes

path = '/home/juan/Insync/juan.zamora@pucv.cl/Google Drive/Research - Multiview and Collaborative Clustering/code/handwritten_tezt/similarity_matrices_after_merging'

In [3]:
from scipy.sparse import csc_matrix, lil_matrix
from itertools import product
#from spectral_embedding import *
from sklearn.cluster import MiniBatchKMeans
from sklearn.manifold import SpectralEmbedding, spectral_embedding
from clustering_evaluation import computeEvaluationMeasures, MEASURES
from sklearn.metrics.pairwise import pairwise_kernels

import scipy.linalg.blas

from scipy.sparse.linalg import norm
import scipy.sparse as sp

from scipy.sparse.linalg import eigsh
from scipy.sparse.csgraph import laplacian as csgraph_laplacian

def getSimClus(D):
    """
    Computes a matrix that contains the intersection of items
    D is a csc_matrix with n rows denoting data points and k columns denoting clusters.
    D_ij denotes the membership score of data point i in cluster j.
    """
    D_asCol = csc_matrix(D) # more efficient for making column slices

    # create a kxk matrix with zeros
    k = D.shape[1]
    simClus = np.zeros((k,k))
    # compare each pair of clusters from D
    for i in range(k):
        sClust_i = set(D_asCol[:,i].nonzero()[0])
        for j in range(i+1,k):
            sClust_j = set(D_asCol[:,j].nonzero()[0])
            simClus[i,j] = len(sClust_i & sClust_j) / len(sClust_i | sClust_j)
            simClus[j,i] = simClus[i,j]
    return simClus


def compute_aSRS(D, DC=1.0):
    # aim: compute the approximate SimRankSimilarity matrix (aSRS).
    # compute for each pair of data points i,j the total sum of simclus values 
    n = D.shape[0]
    
    simClus = getSimClus(D)

    aSRS = np.zeros((n,n))
    for i in range(n):
        nbors_i = D[i].nonzero()[1]
        n_nbors_i = len(nbors_i)
        for j in range(i+1, n):
            nbors_j = D[j].nonzero()[1]
            n_nbors_j = len(nbors_j)
            aSRS[i,j] = (DC/(n_nbors_i*n_nbors_j))*np.sum([simClus[ca,cb] for ca, cb in list(product(nbors_i, nbors_j))])
            aSRS[j,i] = aSRS[i,j]
    return aSRS

In [6]:
def perform_evaluate_runs_part1(dsDirPath, nruns=10, threshold=0.0, embDim=4, kernelFn='rbf'):
    trueLblsFname = "{0}/{1}/{2}".format(path, dsDirPath, 'labels.true')
    trueL = getLabelsFromFile(trueLblsFname)
    postFixesDS = detectPostfixesInDir(path, dsDirPath)

    dsresults = []
    header = ['seed']
    header.extend(MEASURES)
    np.random.seed(87009)
    n_clusters = np.unique(trueL).shape[0]
    runSeed = np.random.randint(0, high=2e8)
    for runPostfix in tqdm(postFixesDS[:1]):#tqdm(postFixesDS):
        #print("{0}/{1}{2}_csr.npy".format(dsDirPath, 'simmatrix_DS_', runPostfix))        
        simFname = "{0}/{1}/{2}{3}_csr.npy".format(path, dsDirPath, 'simmatrix_DS_', runPostfix)
        D = _loadSparseCSRFromDisk(simFname)

        #aSRS = compute_aSRS(D)
        #outSRSFn = "{0}/{1}/{2}{3}.npy".format(path, dsDirPath, 'aSRS_DS_', runPostfix)
        #np.save(outSRSFn, aSRS)
        ##aSRS = np.load(outSRSFn)
        
        #pK = pairwise_kernels(D, metric=kernelFn)
        ## sparse version
        nrmDiag = sp.diags(1/norm(D, axis=1))
        nrmDiag = nrmDiag.dot(D) #heavy
        convert_dense_csv_to_csr_fmt("{0}{1}_csr.npy".format('dotmatrix_DS_', runPostfix), nrmDiag.dot(nrmDiag.T))
        
        
        
        #A = csr_matrix(scipy.linalg.blas.dgemm(alpha=1.0, a=A, b=A, trans_b=True))        

In [4]:
nruns=10
embDim=5

A = _loadSparseCSRFromDisk('dotmatrix_DS_Nov272020.011936_csr_csr.npy')

In [10]:
100*A.nnz/np.product(A.shape)

99.07220133333334

In [9]:
A[0:5,0:5][A[0:5,0:5].nonzero()]

matrix([[0.98162333, 0.98419298, 0.98665904, 0.95655749, 1.        ,
         0.96196538, 0.96080638, 0.96181095, 1.        , 0.95655749,
         0.99906182, 0.96181095, 0.98665904, 0.99903456, 1.        ,
         0.99738909, 1.        , 0.99906182, 0.96196538, 0.98419298,
         0.99738909, 0.98162333, 0.96080638, 0.99903456, 1.        ]])

In [5]:
A.data.shape

(891649812,)

In [6]:
A.indices.shape

(891649812,)

In [7]:
A.indptr.shape

(30001,)

In [5]:
filtered_dotM = sp.csr_matrix(A[A.nonzero()] > 0.9979201675637465)

MemoryError: Unable to allocate 3.32 GiB for an array with shape (891649812,) and data type int32

In [21]:
np.percentile(A.data, 25)

0.9222004352476311

In [10]:
np.percentile(A.data, 80)

0.9979201675637465

In [13]:
from megaman.embedding import spectral_embedding, SpectralEmbedding
from megaman.geometry import Geometry

radius = 5
adjacency_method = 'cyflann'
adjacency_kwds = {'radius':radius} # ignore distances above this radius
affinity_method = 'gaussian'
affinity_kwds = {'radius':radius} # A = exp(-||x - y||/radius^2)
laplacian_method = 'geometric'
laplacian_kwds = {'scaling_epps':radius} # scaling ensures convergence to Laplace-Beltrami operator

geom  = Geometry(adjacency_method=adjacency_method, adjacency_kwds=adjacency_kwds,
                 affinity_method=affinity_method, affinity_kwds=affinity_kwds,
                 laplacian_method=laplacian_method, laplacian_kwds=laplacian_kwds)

In [ ]:
spectral_embedding(acc)

In [14]:
spectral = SpectralEmbedding(n_components=5, eigen_solver='arpack', geom=geom)
embed_spectral = spectral.fit_transform(A.toarray())

MemoryError: Unable to allocate 6.71 GiB for an array with shape (900000000,) and data type float64

In [8]:
embed_spectral

array([[ 1.74881757e-02,  1.02139893e-01,  3.08825753e-02,
        -6.13210481e-02,  2.10219705e-02],
       [ 6.62015550e-02,  6.56135949e-02, -1.19013128e-01,
        -5.17758891e-02, -4.12430387e-02],
       [ 8.27119730e-02,  3.21363162e-02,  2.17106462e-03,
         6.11071568e-02, -1.47006955e-02],
       [ 2.06302536e-01, -6.72965193e-02,  7.14296284e-02,
        -3.13936801e-02, -1.66171624e-01],
       [ 1.93343563e-01, -1.11439747e-01,  1.77459713e-01,
        -4.13820386e-02, -2.04701740e-01],
       [ 9.49120419e-02, -7.36004531e-02,  1.07008162e-01,
         6.47614900e-02,  1.89461394e-01],
       [-2.41788801e-02,  2.81799665e-02,  1.16680343e-01,
        -2.99564210e-02,  2.25491036e-01],
       [ 1.08044796e-01,  4.63797587e-02, -2.24759206e-01,
        -2.49226715e-02, -3.07709983e-02],
       [ 1.24136632e-01,  7.66620227e-02, -2.29966056e-01,
         1.76221910e-03, -2.62220952e-05],
       [ 3.25773883e-02, -4.94612574e-02, -1.46039171e-01,
         1.62335478e-02

In [ ]:
embedding = spectral_embedding(A, n_components=embDim, norm_laplacian=True, drop_first=True, eigen_solver='arpack')

In [7]:
perform_evaluate_runs_part1('NusWide', embDim=5, kernelFn='cosine') # no important difference with the default thresold

  0%|          | 0/1 [00:00<?, ?it/s]

dotmatrix_DS_Nov272020.011936_csr_csr.npy


100%|██████████| 1/1 [00:36<00:00, 36.92s/it]


In [4]:
def perform_evaluate_runs(dsDirPath, nruns=10, threshold=0.0, embDim=4, kernelFn='rbf'):
    trueLblsFname = "{0}/{1}/{2}".format(path, dsDirPath, 'labels.true')
    trueL = getLabelsFromFile(trueLblsFname)
    postFixesDS = detectPostfixesInDir(path, dsDirPath)

    dsresults = []
    header = ['seed']
    header.extend(MEASURES)
    np.random.seed(87009)
    n_clusters = np.unique(trueL).shape[0]
    runSeed = np.random.randint(0, high=2e8)
    for runPostfix in tqdm(postFixesDS[0:1]):#tqdm(postFixesDS):
        #print("{0}/{1}{2}_csr.npy".format(dsDirPath, 'simmatrix_DS_', runPostfix))        
        simFname = "{0}/{1}/{2}{3}_csr.npy".format(path, dsDirPath, 'simmatrix_DS_', runPostfix)
        D = _loadSparseCSRFromDisk(simFname)

        #aSRS = compute_aSRS(D)
        #outSRSFn = "{0}/{1}/{2}{3}.npy".format(path, dsDirPath, 'aSRS_DS_', runPostfix)
        #np.save(outSRSFn, aSRS)
        ##aSRS = np.load(outSRSFn)
        
        #pK = pairwise_kernels(D, metric=kernelFn)
        ## sparse version
        nrmDiag = sp.diags(1/norm(D, axis=1))
        bnrm = nrmDiag.dot(D)
        A = bnrm.toarray()
        del D
        del nrmDiag
        del bnrm
        A = scipy.linalg.blas.dgemm(alpha=1.0, a=A, b=A, trans_b=True)
      
        # builds the embedded vectors
        embedding = spectral_embedding(A, n_components=embDim, norm_laplacian=True, drop_first=True, eigen_solver='arpack')
        
        #spEmbMod = SpectralEmbedding(n_components=20, affinity='nearest_neighbors', n_jobs=4, eigen_solver='arpack', random_state=runSeed)        
        #embedding = spEmbMod.fit_transform(D.toarray())
        
        #outEmbFn = "{0}/{1}/{2}{3}.npy".format(path, dsDirPath, 'embedding_DS_', runPostfix)
        #np.save(outEmbFn, embedding)
        #embedding = np.load(outEmbFn)
        
        # old, not working!
        #spectralEmb = SpectralEmbedding(embedding_dimension = 30, node_weights = 'degree')
        #spectralEmb.fit(aSRS)
        #embedding = spectralEmb.embedding_
        
        runDsResults = [] # a single run passes through all the npy matrices with the same seed        
        for run in tqdm(range(nruns)): # the partitioning method has a stochastic part, hence several runs are needed.
            kmeans = MiniBatchKMeans(n_clusters, random_state=np.random.randint(0, high=2e8))
            kmeans.fit(embedding)
            pred_labels = list(kmeans.labels_)

            perf = computeEvaluationMeasures(trueL, pred_labels)
            runDsResults.append([perf[m] for m in MEASURES])

        runRow = [runSeed]
        runRow.extend(list(np.mean(runDsResults, axis=0)))
        dsresults.append( runRow ) # average

    dsresults.append( ['Average'] + list(np.mean(dsresults, axis=0)[1:]) )
    print(tabulate(dsresults, headers=header, tablefmt='github', floatfmt='.6f', showindex=True))    

**Build the SRS matrix from a bipartite graph adjacency matrix**

In [6]:
%%time
perform_evaluate_runs('BBC-seg4', embDim=4, kernelFn='cosine')

100%|██████████| 10/10 [00:03<00:00,  3.12it/s]

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.183388 | 0.932847 | 0.305188 | 0.304526 | 0.807615 | 0.309141 | 0.304526 | 0.835237 |
|  1 | 102758375 | 0.181714 | 0.933285 | 0.098665 | 0.100000 | 0.809284 | 0.101374 | 0.100000 | 0.836075 |
|  2 | 102758375 | 0.183388 | 0.932847 | 0.120351 | 0.122336 | 0.807615 | 0.120372 | 0.122336 | 0.835237 |
|  3 | 102758375 | 0.184038 | 0.932555 | 0.139719 | 0.142774 | 0.806961 | 0.141232 | 0.142774 | 0.834496 |
|  4 | 102758375 | 0.182830 | 0.932993 | 0.116874 | 0.117664 | 0.808172 | 0.120214 | 0.117664 | 0.835517 |
|  5 | 102758375 | 0.183388 | 0.932847 | 0.161820 | 0.164088 | 0.807615 | 0.162908 | 0.164088 | 0.835237 |
|  6 | 102758375 | 0.182830 | 0.932993 | 0.090352 | 0.090219 | 0.808172 | 0.094477 | 0.090219 | 0.835517 |
|  7 | 102758375 | 0.182830 | 0.93299

dim:4, cosine

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.183388 | 0.932847 | 0.305188 | 0.304526 | 0.807615 | 0.309141 | 0.304526 | 0.835237 |
|  1 | 102758375 | 0.181714 | 0.933285 | 0.098665 | 0.100000 | 0.809284 | 0.101374 | 0.100000 | 0.836075 |
|  2 | 102758375 | 0.183388 | 0.932847 | 0.120351 | 0.122336 | 0.807615 | 0.120372 | 0.122336 | 0.835237 |
|  3 | 102758375 | 0.184038 | 0.932555 | 0.139719 | 0.142774 | 0.806961 | 0.141232 | 0.142774 | 0.834496 |
|  4 | 102758375 | 0.182830 | 0.932993 | 0.116874 | 0.117664 | 0.808172 | 0.120214 | 0.117664 | 0.835517 |
|  5 | 102758375 | 0.183388 | 0.932847 | 0.161820 | 0.164088 | 0.807615 | 0.162908 | 0.164088 | 0.835237 |
|  6 | 102758375 | 0.182830 | 0.932993 | 0.090352 | 0.090219 | 0.808172 | 0.094477 | 0.090219 | 0.835517 |
|  7 | 102758375 | 0.182830 | 0.932993 | 0.138518 | 0.140146 | 0.808172 | 0.140495 | 0.140146 | 0.835517 |
|  8 | 102758375 | 0.181985 | 0.933139 | 0.171865 | 0.174891 | 0.808953 | 0.173101 | 0.174891 | 0.835747 |
|  9 | 102758375 | 0.184038 | 0.932555 | 0.208822 | 0.211095 | 0.806961 | 0.210462 | 0.211095 | 0.834496 |
| 10 | Average   | 0.183043 | 0.932905 | 0.155217 | 0.156774 | 0.807952 | 0.157378 | 0.156774 | 0.835308 |
CPU times: user 8.88 s, sys: 4.01 s, total: 12.9 s
Wall time: 2.95 s



|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.268472 | 0.894161 | 0.149339 | 0.155328 | 0.721357 | 0.148146 | 0.155328 | 0.755605 |
|  1 | 102758375 | 0.268042 | 0.894307 | 0.094929 | 0.097664 | 0.721754 | 0.098474 | 0.097664 | 0.755752 |
|  2 | 102758375 | 0.269831 | 0.893431 | 0.118097 | 0.124380 | 0.720049 | 0.117356 | 0.124380 | 0.754138 |
|  3 | 102758375 | 0.269305 | 0.893723 | 0.210728 | 0.215182 | 0.720558 | 0.211137 | 0.215182 | 0.754575 |
|  4 | 102758375 | 0.266446 | 0.895182 | 0.153666 | 0.154599 | 0.723294 | 0.158414 | 0.154599 | 0.757712 |
|  5 | 102758375 | 0.267779 | 0.894453 | 0.135532 | 0.139416 | 0.722008 | 0.136482 | 0.139416 | 0.755971 |
|  6 | 102758375 | 0.268304 | 0.894307 | 0.133509 | 0.138832 | 0.721531 | 0.132912 | 0.138832 | 0.755996 |
|  7 | 102758375 | 0.267542 | 0.894599 | 0.146864 | 0.150949 | 0.722240 | 0.148295 | 0.150949 | 0.756463 |
|  8 | 102758375 | 0.272487 | 0.890219 | 0.300385 | 0.306861 | 0.716485 | 0.301102 | 0.306861 | 0.749963 |
|  9 | 102758375 | 0.268972 | 0.893869 | 0.039151 | 0.043942 | 0.720872 | 0.040679 | 0.043942 | 0.754894 |
| 10 | Average   | 0.268718 | 0.893825 | 0.148220 | 0.152715 | 0.721015 | 0.149300 | 0.152715 | 0.755107 |

CPU times: user 4min 56s, sys: 2.46 s, total: 4min 58s
Wall time: 4min 46s

In [13]:
%%time
perform_evaluate_runs('Caltech', embDim=8, kernelFn='cosine')

100%|██████████| 10/10 [00:11<00:00,  1.10s/it]

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.428383 | 0.659640 | 0.043620 | 0.044007 | 0.436066 | 0.073938 | 0.044007 | 0.303957 |
|  1 | 102758375 | 0.427989 | 0.662825 | 0.065009 | 0.058256 | 0.436709 | 0.097686 | 0.058256 | 0.318946 |
|  2 | 102758375 | 0.418265 | 0.652850 | 0.113960 | 0.106873 | 0.450076 | 0.199417 | 0.106873 | 0.346506 |
|  3 | 102758375 | 0.419574 | 0.652096 | 0.068019 | 0.066345 | 0.444782 | 0.104811 | 0.066345 | 0.336888 |
|  4 | 102758375 | 0.429605 | 0.658466 | 0.051974 | 0.055532 | 0.436864 | 0.076387 | 0.055532 | 0.318954 |
|  5 | 102758375 | 0.428818 | 0.664878 | 0.062528 | 0.059137 | 0.439885 | 0.154630 | 0.059137 | 0.314035 |
|  6 | 102758375 | 0.425140 | 0.645013 | 0.048215 | 0.044593 | 0.439948 | 0.109057 | 0.044593 | 0.335990 |
|  7 | 102758375 | 0.428611 | 0.66311

dim:8, cosine

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.428383 | 0.659640 | 0.043620 | 0.044007 | 0.436066 | 0.073938 | 0.044007 | 0.303957 |
|  1 | 102758375 | 0.427989 | 0.662825 | 0.065009 | 0.058256 | 0.436709 | 0.097686 | 0.058256 | 0.318946 |
|  2 | 102758375 | 0.418265 | 0.652850 | 0.113960 | 0.106873 | 0.450076 | 0.199417 | 0.106873 | 0.346506 |
|  3 | 102758375 | 0.419574 | 0.652096 | 0.068019 | 0.066345 | 0.444782 | 0.104811 | 0.066345 | 0.336888 |
|  4 | 102758375 | 0.429605 | 0.658466 | 0.051974 | 0.055532 | 0.436864 | 0.076387 | 0.055532 | 0.318954 |
|  5 | 102758375 | 0.428818 | 0.664878 | 0.062528 | 0.059137 | 0.439885 | 0.154630 | 0.059137 | 0.314035 |
|  6 | 102758375 | 0.425140 | 0.645013 | 0.048215 | 0.044593 | 0.439948 | 0.109057 | 0.044593 | 0.335990 |
|  7 | 102758375 | 0.428611 | 0.663118 | 0.057425 | 0.056370 | 0.439868 | 0.097537 | 0.056370 | 0.314683 |
|  8 | 102758375 | 0.437024 | 0.656035 | 0.049690 | 0.044677 | 0.430078 | 0.095058 | 0.044677 | 0.333822 |
|  9 | 102758375 | 0.415802 | 0.653353 | 0.033990 | 0.034996 | 0.449886 | 0.052390 | 0.034996 | 0.351205 |
| 10 | Average   | 0.425921 | 0.656827 | 0.059443 | 0.057079 | 0.440416 | 0.106091 | 0.057079 | 0.327498 |
CPU times: user 31.9 s, sys: 9.97 s, total: 41.9 s
Wall time: 11 s


dim:5 , cosine

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.429888 | 0.648910 | 0.052559 | 0.052096 | 0.425421 | 0.078130 | 0.052096 | 0.322740 |
|  1 | 102758375 | 0.430596 | 0.649581 | 0.050770 | 0.046773 | 0.423121 | 0.094186 | 0.046773 | 0.316401 |
|  2 | 102758375 | 0.405382 | 0.659514 | 0.094435 | 0.085247 | 0.450282 | 0.159397 | 0.085247 | 0.354355 |
|  3 | 102758375 | 0.407962 | 0.655448 | 0.073609 | 0.064920 | 0.448392 | 0.125081 | 0.064920 | 0.360443 |
|  4 | 102758375 | 0.430999 | 0.648701 | 0.023385 | 0.027494 | 0.420953 | 0.038343 | 0.027494 | 0.309670 |
|  5 | 102758375 | 0.431033 | 0.650838 | 0.057918 | 0.048868 | 0.422462 | 0.109296 | 0.048868 | 0.314567 |
|  6 | 102758375 | 0.409791 | 0.655029 | 0.064198 | 0.054862 | 0.445235 | 0.138087 | 0.054862 | 0.348076 |
|  7 | 102758375 | 0.429922 | 0.648868 | 0.058080 | 0.049162 | 0.425611 | 0.120052 | 0.049162 | 0.328405 |
|  8 | 102758375 | 0.422801 | 0.661065 | 0.071160 | 0.065423 | 0.433804 | 0.118065 | 0.065423 | 0.350798 |
|  9 | 102758375 | 0.407643 | 0.656329 | 0.065358 | 0.062531 | 0.448731 | 0.142431 | 0.062531 | 0.349878 |
| 10 | Average   | 0.420602 | 0.653428 | 0.061147 | 0.055738 | 0.434401 | 0.112307 | 0.055738 | 0.335533 |
CPU times: user 31.4 s, sys: 9.99 s, total: 41.4 s
Wall time: 10.9 s

dim:3, cosine

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.429080 | 0.649246 | 0.030382 | 0.031140 | 0.425903 | 0.045765 | 0.031140 | 0.350819 |
|  1 | 102758375 | 0.430828 | 0.649329 | 0.034217 | 0.034367 | 0.423641 | 0.051573 | 0.034367 | 0.345043 |
|  2 | 102758375 | 0.414870 | 0.643713 | 0.093608 | 0.080134 | 0.433554 | 0.189920 | 0.080134 | 0.312642 |
|  3 | 102758375 | 0.415314 | 0.643588 | 0.055367 | 0.047946 | 0.432987 | 0.114854 | 0.047946 | 0.312657 |
|  4 | 102758375 | 0.427223 | 0.650168 | 0.074751 | 0.065130 | 0.427829 | 0.117880 | 0.065130 | 0.352501 |
|  5 | 102758375 | 0.429397 | 0.648156 | 0.040202 | 0.037427 | 0.422792 | 0.073292 | 0.037427 | 0.339596 |
|  6 | 102758375 | 0.417945 | 0.643588 | 0.058441 | 0.053646 | 0.428766 | 0.116574 | 0.053646 | 0.299925 |
|  7 | 102758375 | 0.429430 | 0.650754 | 0.049732 | 0.045851 | 0.427227 | 0.093415 | 0.045851 | 0.357674 |
|  8 | 102758375 | 0.428885 | 0.648659 | 0.057716 | 0.059262 | 0.420952 | 0.088058 | 0.059262 | 0.325333 |
|  9 | 102758375 | 0.414440 | 0.643965 | 0.068801 | 0.054694 | 0.432413 | 0.199044 | 0.054694 | 0.310104 |
| 10 | Average   | 0.423741 | 0.647117 | 0.056322 | 0.050960 | 0.427606 | 0.109038 | 0.050960 | 0.330629 |
CPU times: user 31.9 s, sys: 9.51 s, total: 41.4 s
Wall time: 11.2 s


dim : 3

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.445440 | 0.643336 | 0.026973 | 0.027033 | 0.414287 | 0.055858 | 0.027033 | 0.342926 |
|  1 | 102758375 | 0.444238 | 0.643336 | 0.045111 | 0.041408 | 0.416247 | 0.070447 | 0.041408 | 0.348293 |
|  2 | 102758375 | 0.424884 | 0.648198 | 0.077482 | 0.071961 | 0.426085 | 0.128635 | 0.071961 | 0.304730 |
|  3 | 102758375 | 0.423435 | 0.648198 | 0.077158 | 0.065339 | 0.426261 | 0.186978 | 0.065339 | 0.301516 |
|  4 | 102758375 | 0.443114 | 0.643168 | 0.046479 | 0.041869 | 0.415925 | 0.073534 | 0.041869 | 0.346467 |
|  5 | 102758375 | 0.441940 | 0.644216 | 0.027960 | 0.033445 | 0.415597 | 0.042074 | 0.033445 | 0.339148 |
|  6 | 102758375 | 0.422699 | 0.645306 | 0.045728 | 0.044216 | 0.426639 | 0.077841 | 0.044216 | 0.304276 |
|  7 | 102758375 | 0.443025 | 0.644258 | 0.057770 | 0.052054 | 0.419893 | 0.104900 | 0.052054 | 0.376782 |
|  8 | 102758375 | 0.444823 | 0.644593 | 0.035458 | 0.036421 | 0.411878 | 0.068847 | 0.036421 | 0.349812 |
|  9 | 102758375 | 0.422160 | 0.648617 | 0.055328 | 0.041953 | 0.428452 | 0.199043 | 0.041953 | 0.308407 |
| 10 | Average   | 0.435576 | 0.645323 | 0.049545 | 0.045570 | 0.420126 | 0.100816 | 0.045570 | 0.332236 |

CPU times: user 1min 46s, sys: 57.2 s, total: 2min 44s
Wall time: 28.2 s

In [38]:
%%time
perform_evaluate_runs('handwritten', embDim=6, kernelFn='cosine') # no important difference with the default thresold

100%|██████████| 10/10 [00:07<00:00,  1.40it/s]

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.281818 | 0.761050 | 0.095272 | 0.099400 | 0.730629 | 0.101884 | 0.099400 | 0.627191 |
|  1 | 102758375 | 0.276767 | 0.770700 | 0.069029 | 0.077700 | 0.734134 | 0.066716 | 0.077700 | 0.637387 |
|  2 | 102758375 | 0.277524 | 0.764200 | 0.138228 | 0.145250 | 0.733049 | 0.142899 | 0.145250 | 0.636140 |
|  3 | 102758375 | 0.278512 | 0.770100 | 0.119809 | 0.123950 | 0.732739 | 0.125409 | 0.123950 | 0.632471 |
|  4 | 102758375 | 0.278240 | 0.764150 | 0.095165 | 0.102700 | 0.732373 | 0.098505 | 0.102700 | 0.630341 |
|  5 | 102758375 | 0.284441 | 0.763200 | 0.074636 | 0.077600 | 0.729158 | 0.077518 | 0.077600 | 0.619431 |
|  6 | 102758375 | 0.275399 | 0.774150 | 0.118619 | 0.124000 | 0.734071 | 0.118853 | 0.124000 | 0.640756 |
|  7 | 102758375 | 0.302031 | 0.74215

dim:6, cosine 

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.281818 | 0.761050 | 0.095272 | 0.099400 | 0.730629 | 0.101884 | 0.099400 | 0.627191 |
|  1 | 102758375 | 0.276767 | 0.770700 | 0.069029 | 0.077700 | 0.734134 | 0.066716 | 0.077700 | 0.637387 |
|  2 | 102758375 | 0.277524 | 0.764200 | 0.138228 | 0.145250 | 0.733049 | 0.142899 | 0.145250 | 0.636140 |
|  3 | 102758375 | 0.278512 | 0.770100 | 0.119809 | 0.123950 | 0.732739 | 0.125409 | 0.123950 | 0.632471 |
|  4 | 102758375 | 0.278240 | 0.764150 | 0.095165 | 0.102700 | 0.732373 | 0.098505 | 0.102700 | 0.630341 |
|  5 | 102758375 | 0.284441 | 0.763200 | 0.074636 | 0.077600 | 0.729158 | 0.077518 | 0.077600 | 0.619431 |
|  6 | 102758375 | 0.275399 | 0.774150 | 0.118619 | 0.124000 | 0.734071 | 0.118853 | 0.124000 | 0.640756 |
|  7 | 102758375 | 0.302031 | 0.742150 | 0.116881 | 0.123550 | 0.713664 | 0.121963 | 0.123550 | 0.605184 |
|  8 | 102758375 | 0.277821 | 0.765100 | 0.100481 | 0.112100 | 0.732227 | 0.098190 | 0.112100 | 0.634336 |
|  9 | 102758375 | 0.279416 | 0.768800 | 0.075536 | 0.075800 | 0.732158 | 0.077696 | 0.075800 | 0.627847 |
| 10 | Average   | 0.281197 | 0.764360 | 0.100366 | 0.106205 | 0.730420 | 0.102963 | 0.106205 | 0.629108 |
CPU times: user 21.4 s, sys: 7.06 s, total: 28.5 s
Wall time: 7.18 s

dim : 6

|    | seed      |        E |        P |       F1 |      ACC |      NMI |     PREC |      REC |      ARI |
|----|-----------|----------|----------|----------|----------|----------|----------|----------|----------|
|  0 | 102758375 | 0.290102 | 0.763400 | 0.121379 | 0.121450 | 0.716668 | 0.125602 | 0.121450 | 0.634751 |
|  1 | 102758375 | 0.289881 | 0.763200 | 0.073071 | 0.075850 | 0.716801 | 0.073154 | 0.075850 | 0.630876 |
|  2 | 102758375 | 0.290499 | 0.761100 | 0.132062 | 0.133500 | 0.715696 | 0.134442 | 0.133500 | 0.634909 |
|  3 | 102758375 | 0.288695 | 0.764150 | 0.095413 | 0.098150 | 0.717387 | 0.095284 | 0.098150 | 0.634717 |
|  4 | 102758375 | 0.293103 | 0.758750 | 0.124142 | 0.126900 | 0.715118 | 0.127762 | 0.126900 | 0.627984 |
|  5 | 102758375 | 0.289417 | 0.762350 | 0.100434 | 0.099700 | 0.717925 | 0.106166 | 0.099700 | 0.630034 |
|  6 | 102758375 | 0.289903 | 0.763300 | 0.145624 | 0.148700 | 0.716421 | 0.149754 | 0.148700 | 0.632362 |
|  7 | 102758375 | 0.337331 | 0.688300 | 0.098746 | 0.100600 | 0.673209 | 0.105153 | 0.100600 | 0.583970 |
|  8 | 102758375 | 0.290753 | 0.762650 | 0.080585 | 0.080400 | 0.717223 | 0.086493 | 0.080400 | 0.628180 |
|  9 | 102758375 | 0.290330 | 0.762950 | 0.071886 | 0.069750 | 0.716495 | 0.078104 | 0.069750 | 0.634831 |
| 10 | Average   | 0.295001 | 0.755015 | 0.104334 | 0.105500 | 0.712294 | 0.108191 | 0.105500 | 0.627261 |

CPU times: user 56.9 s, sys: 31.1 s, total: 1min 28s
Wall time: 15 s

In [5]:
perform_evaluate_runs('NusWide', embDim=5, kernelFn='cosine') # no important difference with the default thresold

  0%|          | 0/1 [00:09<?, ?it/s]


MemoryError: Unable to allocate 6.71 GiB for an array with shape (30000, 30000) and data type float64

---

In [29]:
dsDirPath = 'BBC-seg4'
simFname = "{0}/{1}/{2}".format(path, dsDirPath, 'simmatrix_DS_Nov262020.182009_csr.npy')
trueLblsFname = "{0}/{1}/{2}".format(path, dsDirPath, 'labels.true')
trueL = getLabelsFromFile(trueLblsFname)
D = _loadSparseCSRFromDisk(simFname)
D

<685x15 sparse matrix of type '<class 'numpy.float64'>'
	with 1453 stored elements in Compressed Sparse Row format>

In [30]:
simClus = getSimClus(D)
simClus.shape

(15, 15)

In [31]:
%%time
aSRS = compute_aSRS(D)

CPU times: user 28.7 s, sys: 83.2 ms, total: 28.8 s
Wall time: 28.9 s


Obtaining the graph embedding

In [116]:
from spectral_embedding import *

spectralEmb = SpectralEmbedding(embedding_dimension = 30, node_weights = 'degree')
spectralEmb.fit(aSRS)

embedding = spectralEmb.embedding_
embedding.shape

from sklearn.cluster import KMeans, MiniBatchKMeans

n_clusters = 5

kmeans = MiniBatchKMeans(n_clusters)
kmeans.fit(embedding)
pred_labels = list(kmeans.labels_)
np.unique(pred_labels, return_counts=True)

from clustering_evaluation import computeEvaluationMeasures, MEASURES

perf = computeEvaluationMeasures(trueL, pred_labels) 
perf

In [56]:
from sklearn.manifold import SpectralEmbedding, spectral_embedding, LocallyLinearEmbedding
from clustering_evaluation import computeEvaluationMeasures, MEASURES
from sklearn.cluster import MiniBatchKMeans

In [33]:
%%time
# builds the embedded vectors
embedding = SpectralEmbedding(n_components=20, affinity='precomputed', n_jobs=4)
D_transformed = embedding.fit_transform(csr_matrix(aSRS))
D_transformed.shape

CPU times: user 2min 10s, sys: 1min 59s, total: 4min 9s
Wall time: 33.7 s


(685, 20)

In [23]:
# clustering the embedded vectors
n_clusters = 5

kmeans = MiniBatchKMeans(n_clusters)
kmeans.fit(D_transformed)
pred_labels = list(kmeans.labels_)
np.unique(pred_labels, return_counts=True)

# evaluates
perf = computeEvaluationMeasures(trueL, pred_labels) 
perf

{'E': 0.2698309799979815,
 'P': 0.8934306569343065,
 'F1': 0.034864843556005734,
 'ACC': 0.043795620437956206,
 'NMI': 0.720048774881849,
 'PREC': 0.031503791390491524,
 'REC': 0.043795620437956206,
 'ARI': 0.7541376286640216}

In [72]:
D_transformed = spectral_embedding(aSRS, n_components=4)

In [73]:
# clustering the embedded vectors
n_clusters = 5
results = []
for i in range(10):
    kmeans = MiniBatchKMeans(n_clusters)
    kmeans.fit(D_transformed)
    pred_labels = list(kmeans.labels_)
    np.unique(pred_labels, return_counts=True)

    # evaluates
    perf = computeEvaluationMeasures(trueL, pred_labels)
    results.append([perf[m] for m in MEASURES])    

results.append(list(np.mean(results, axis=0)))

print(tabulate(results, headers=MEASURES, tablefmt='github'))

|        E |        P |        F1 |       ACC |      NMI |      PREC |       REC |      ARI |
|----------|----------|-----------|-----------|----------|-----------|-----------|----------|
| 0.269831 | 0.893431 | 0.0272362 | 0.0364964 | 0.720049 | 0.0247993 | 0.0364964 | 0.754138 |
| 0.269831 | 0.893431 | 0.0348648 | 0.0437956 | 0.720049 | 0.0315038 | 0.0437956 | 0.754138 |
| 0.269831 | 0.893431 | 0.0183744 | 0.0189781 | 0.720049 | 0.0196187 | 0.0189781 | 0.754138 |
| 0.269831 | 0.893431 | 0.119847  | 0.124088  | 0.720049 | 0.116915  | 0.124088  | 0.754138 |
| 0.269831 | 0.893431 | 0.395171  | 0.386861  | 0.720049 | 0.407427  | 0.386861  | 0.754138 |
| 0.422587 | 0.729927 | 0.180346  | 0.157664  | 0.602769 | 0.312192  | 0.157664  | 0.48873  |
| 0.269831 | 0.893431 | 0.0269503 | 0.0321168 | 0.720049 | 0.0255184 | 0.0321168 | 0.754138 |
| 0.269831 | 0.893431 | 0.572585  | 0.581022  | 0.720049 | 0.568539  | 0.581022  | 0.754138 |
| 0.264828 | 0.89635  | 0.242243  | 0.240876  | 0.724914 | 0

array([0.40443594, 0.74306569, 0.20445587, 0.23781022, 0.6085907 ,
       0.2044238 , 0.23781022, 0.53578728])

In [10]:
Sp = np.abs(np.random.randn(4,3))
Sp

array([[0.80471619, 0.22661069, 0.70333301],
       [1.21995249, 0.7524992 , 0.60835408],
       [0.81638102, 0.21234497, 1.29083036],
       [1.21960049, 1.22693506, 1.15690697]])

In [12]:
from sklearn.metrics.pairwise import pairwise_kernels

pairwise_kernels(Sp, metric='cosine')

array([[1.        , 0.92883101, 0.95742643, 0.91179706],
       [0.92883101, 1.        , 0.80838389, 0.96121347],
       [0.95742643, 0.80838389, 1.        , 0.85678013],
       [0.91179706, 0.96121347, 0.85678013, 1.        ]])

In [13]:
pairwise_kernels(Sp, metric='rbf')

array([[1.        , 0.85841556, 0.89121976, 0.63159786],
       [0.85841556, 1.        , 0.7357952 , 0.83917731],
       [0.89121976, 0.7357952 , 1.        , 0.66810842],
       [0.63159786, 0.83917731, 0.66810842, 1.        ]])

In [15]:
pairwise_kernels(Sp, metric='laplacian')

array([[1.        , 0.70795984, 0.81507433, 0.53637189],
       [0.70795984, 1.        , 0.58154477, 0.71097815],
       [0.81507433, 0.58154477, 1.        , 0.59616112],
       [0.53637189, 0.71097815, 0.59616112, 1.        ]])

In [16]:
Sp.dot(Sp.T)

array([[1.19359788, 1.5801154 , 1.61295827, 2.07315973],
       [1.5801154 , 2.42463381, 1.94101739, 3.11493138],
       [1.61295827, 1.94101739, 2.37781137, 2.74956282],
       [2.07315973, 3.11493138, 2.74956282, 4.33122872]])